File: web/src/lib/schemas.ts
import { z } from "zod";

export const guardrailFlagSchema = z.object({
  code: z.string(),
  severity: z.enum(["low", "medium", "high", "critical"]),
  message: z.string(),
  metric: z.string(),
  observed_value: z.number(),
  threshold: z.number(),
  policy_ref: z.string()
});

export const guardrailResultSchema = z.object({
  status: z.enum(["pass", "review", "block"]),
  flags: z.array(guardrailFlagSchema)
});

export const regretResultSchema = z.object({
  status: z.enum(["pending", "scored", "unavailable"]),
  score_inr: z.number(),
  best_price_inr: z.number(),
  chosen_reward_inr: z.number(),
  best_reward_inr: z.number(),
  model_version: z.string()
});

export const explanationEvidenceSchema = z.object({
  doc_id: z.string(),
  title: z.string(),
  snippet: z.string(),
  score: z.number(),
  policy_ref: z.string()
});

export const explanationResultSchema = z.object({
  status: z.enum(["grounded", "refused", "unavailable"]),
  summary: z.string(),
  evidence: z.array(explanationEvidenceSchema)
});

export const pricingContextSchema = z
  .object({
    inventory: z.number(),
    competitor_price_inr: z.number(),
    demand_index: z.number()
  })
  .passthrough();

export const anomalyModeSchema = z.enum([
  "none",
  "price_spike",
  "floor_breach",
  "ceiling_breach",
  "drift"
]);

export const decisionEnvelopeSchema = z.object({
  decision_id: z.string(),
  occurred_at: z.string(),
  sku_id: z.string(),
  channel: z.string(),
  region: z.string(),
  context: pricingContextSchema,
  proposed_price_inr: z.number(),
  policy_version: z.string(),
  model_version: z.string(),
  trace_id: z.string(),
  source: z.string(),
  anomaly_mode: anomalyModeSchema,
  guardrail: guardrailResultSchema,
  regret: regretResultSchema,
  explanation: explanationResultSchema
});

export const dashboardFeedSchema = z.object({
  generated_at: z.string(),
  decisions: z.array(decisionEnvelopeSchema)
});

export const decisionListSchema = z.object({
  decisions: z.array(decisionEnvelopeSchema)
});

export type AnomalyMode = z.infer<typeof anomalyModeSchema>;
export type DecisionEnvelope = z.infer<typeof decisionEnvelopeSchema>;
export type DashboardFeed = z.infer<typeof dashboardFeedSchema>;

File: backend/app/contracts/schemas.py
from datetime import datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class AnomalyMode(StrEnum):
    none = "none"
    price_spike = "price_spike"
    floor_breach = "floor_breach"
    ceiling_breach = "ceiling_breach"
    drift = "drift"


class PricingContext(BaseModel):
    model_config = ConfigDict(extra="allow")

    inventory: int = Field(ge=0)
    competitor_price_inr: float = Field(gt=0)
    demand_index: float = Field(ge=0, le=1)


class GuardrailFlag(BaseModel):
    code: str
    severity: Literal["low", "medium", "high", "critical"]
    message: str
    metric: str
    observed_value: float
    threshold: float
    policy_ref: str


class GuardrailResult(BaseModel):
    status: Literal["pass", "review", "block"] = "pass"
    flags: list[GuardrailFlag] = Field(default_factory=list)


class RegretResult(BaseModel):
    status: Literal["pending", "scored", "unavailable"] = "pending"
    score_inr: float = 0
    best_price_inr: float = 0
    chosen_reward_inr: float = 0
    best_reward_inr: float = 0
    model_version: str = "regret-v1"


class ExplanationEvidence(BaseModel):
    doc_id: str
    title: str
    snippet: str
    score: float
    policy_ref: str


class ExplanationResult(BaseModel):
    status: Literal["grounded", "refused", "unavailable"] = "unavailable"
    summary: str = ""
    evidence: list[ExplanationEvidence] = Field(default_factory=list)


class DecisionEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid")

    decision_id: str
    occurred_at: datetime
    sku_id: str
    channel: str
    region: str
    context: PricingContext
    proposed_price_inr: float = Field(gt=0)
    policy_version: str
    model_version: str
    trace_id: str
    source: str
    anomaly_mode: AnomalyMode = AnomalyMode.none
    guardrail: GuardrailResult = Field(default_factory=GuardrailResult)
    regret: RegretResult = Field(default_factory=RegretResult)
    explanation: ExplanationResult = Field(default_factory=ExplanationResult)

    def as_json_dict(self) -> dict[str, Any]:
        return self.model_dump(mode="json")


class SimulationRequest(BaseModel):
    count: int = Field(default=50, ge=1, le=500)
    anomaly_mode: AnomalyMode = AnomalyMode.none


class DecisionListResponse(BaseModel):
    decisions: list[DecisionEnvelope]


class DashboardFeedResponse(BaseModel):
    generated_at: datetime
    decisions: list[DecisionEnvelope]
